buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java

```diff
@@ -0,0 +1,331 @@
+            .collect(Collectors.groupingBy(Version::getMajor, Collectors.toList()));
+
+        if (groupByMajor.isEmpty()) {
+            throw new IllegalArgumentException("Could not parse any versions");
+        }
+
+        currentVersion = getLatestVersionByKey(
+            groupByMajor,
+            groupByMajor.keySet().stream().max(Integer::compareTo)
+                .orElseThrow(() -> new IllegalStateException("Unexpected number of versions in collection"))
+        );
+
+        assertCurrentVersionMatchesParsed(currentVersionProperty);
+
+        assertNoOlderThanTwoMajors();
+
+        markUnreleasedAsSnapshot();
+    }
+
+    private void markUnreleasedAsSnapshot() {
+        getUnreleased()
+            .forEach(unreleased -> {
+                groupByMajor.get(unreleased.getMajor()).remove(unreleased);
+                groupByMajor.get(unreleased.getMajor()).add(
+                    new Version(
+                        unreleased.getMajor(), unreleased.getMinor(), unreleased.getRevision(),
+                        unreleased.getSuffix(), true
+                    )
+                );
+            });
+    }
+
+    private void assertNoOlderThanTwoMajors() {
+        Set<Integer> majors = groupByMajor.keySet();
+        if (majors.size() != 2) {
+            throw new IllegalStateException(
+                "Expected exactly 2 majors in parsed versions but found: " + majors
+            );
+        }
+    }
+
+    private void assertCurrentVersionMatchesParsed(Version currentVersionProperty) {
+        if (currentVersionProperty.equals(currentVersion) == false) {
+            throw new IllegalStateException(
+                "Parsed versions latest version does not match the one configured in build properties. " +
+                    "Parsed latest version is " + currentVersion + " but the build has " +
+                    currentVersionProperty
+            );
+        }
+    }
+
+    public void forPreviousUnreleased(Consumer<UnreleasedVersionDescription> consumer) {
+        getUnreleased().stream()
+            .filter(version -> version.equals(currentVersion) == false)
+            .forEach(version -> consumer.accept(
+                new UnreleasedVersionDescription(
+                    version,
+                    getBranchFor(version),
+                    getGradleProjectNameFor(version)
+                )
+            ));
+    }
+
+    private String getGradleProjectNameFor(Version version) {
+        if (version.equals(currentVersion)) {
+            throw new IllegalArgumentException("The Gradle project to build " + version + " is the current build.");
+        }
+        Map<Integer, List<Version>> releasedMajorGroupedByMinor = getReleasedMajorGroupedByMinor();
+
+        if (version.getRevision() == 0) {
+            if (releasedMajorGroupedByMinor
+                .get(releasedMajorGroupedByMinor.keySet().stream().max(Integer::compareTo).orElse(0))
```

Comment:
The list of versions within a major should be sorted, so we should be able to just look at the last element?

Reply:
It's getting the largest key, i.e. the latest minor. We don't have ordering for the keys.